core/src/main/java/org/eclipse/hono/util/TenantObject.java

```diff
@@ -505,9 +505,22 @@ public static JsonObject newAdapterConfig(final String type, final boolean enabl
      */
     @JsonIgnore
     public long getConnectionsLimit() {
-        return Optional.ofNullable(getProperty(TenantConstants.LIMITS))
-                .map(limits -> getProperty((JsonObject) limits, TenantConstants.MAX_CONNECTIONS,
+        return Optional.ofNullable(getProperty(TenantConstants.RESOURCE_LIMITS))
+                .map(limits -> getProperty((JsonObject) limits, TenantConstants.FIELD_MAX_CONNECTIONS,
                         (Number) TenantConstants.DEFAULT_MAX_CONNECTIONS).longValue())
                 .orElse(TenantConstants.DEFAULT_MAX_CONNECTIONS);
     }
+
+    /**
+     * Gets the data volume limit to be allowed for the tenant per month.
+     *
+     * @return The maximum bytes per month or {@link TenantConstants#DEFAULT_MAX_BYTES_PER_MONTH} if not set.
+     */
+    @JsonIgnore
+    public long getMaxBytesPerMonthLimit() {
```

Comment:
I wonder if this is should better be more generic when it comes to the underlying time frame that is to be considered.
I could imagine to store the *maximum number of bytes* as well as the *number of days* that should be considered when determining if the limit has been reached. WDYT?

Reply:
@sophokles73 It would be good to make it more generic. IMHO just the _number of days_ wouldn't make much sense without knowing _from which point of time_ to calculate the _data consumed_. 
Assume an user would like to limit the data volume to 200 MB every month for a tenant and lets say he/she configures as below.  
```
    "limits": {
      "max-bytes": 209_715_200,
      "max-bytes-period-in-days": 30
    }
```
IMHO calculating the data consumed for the last 30 days wouldn't make any sense. Calculating from the first of the month till the current day is one way. In this case, I can imagine something like the below configuration. If it is quarterly, then calculate from the beginning of this quarter till current date and so for half-yearly and yearly.
```
    "limits": {
      "max-bytes": 209_715_200,
      "max-bytes-period": weekly/monthly/quarterly/half-yearly/yearly
    }
```

As an alternative, the tenant registration date or an explicit user defined date can be used. In the below sample configuration, the time period is set to 90 days. The data consumed every 90 days starting from the device registration date can be used for the limit checks. But the reference device registry implementation doesn't have the information on which date a tenant is registered.
```
    "limits": {
      "max-bytes": 209_715_200,
      "max-bytes-period-in-days": 90
    }
```
WDYT?

Reply:
> IMHO just the number of days wouldn't make much sense without knowing from which point of time to calculate the data consumed.

I agree.

I like the idea of using
```
"max-bytes-period-in-days": 90
```
to indicate the time period over which the limit is enforced. The limit on the data volume is enforced at the tenant level, not the device level. Thus, we only need to know the date at which the tenant has been created or the current plan has been chosen and can the deduce the *current period's starting date* based on that date. We will also need to add properties for tracking these dates to the tenant and/or its *limits* property.

Reply:
@sophokles73 Sorry, my intention was also the tenant registration date and not that of the device. 

>  Thus, we only need to know the date at which the tenant has been created or the current plan has been chosen and can the deduce the _current period's starting date_ based on that date. We will also need to add properties for tracking these dates to the tenant and/or its _limits_ property.

A user can change a tenant plan to another plan, for example from a monthly to a quarterly cycle. I  think it makes more sense to use _when the current plan has been chosen_ instead of the _fixed tenant creation date_ .  I can imagine a property named `plan-start-date` or `effective-date`.  WDYT?

```
    "limits": {
      "max-connections": 100,
      "max-bytes": 209_715_200,
      "max-bytes-period-in-days": 90,
      "effective-date": "2019-04-23"
    }
```

If the above property is not set, the _Protocol Adapters_ consider the `max-bytes-...` config as invalid and refrain from checking the message limits based on data volume. This can be notified through a warn message in logs. WDYT?

Reply:
sounds reasonable to me. I would prefer *effective-since* over *effective-date*, though.
We would then end up with three properties needed to define the data volume limit which to me indicates that we should introduce a new JSON object for it, don't you think?

Reply:
@sophokles73 WDYT about the below config?

   ```
    "resource-limits": {
      "max-connections": 2,
      "max-messages": {
        "bytes-limit": 209_715_200,
        "period-in-days": 90,
        "effective-since": "2019-04-23"
      }
    }
```

Reply:
well, we do not really care about the number of messages, do we? So FMPOV the name should not suggest that we do. What about simply calling it *transferred-data* or *data-volume* and rename the *bytes-limit* to *max-bytes*?

Reply:
My intention was to keep it generic, so that any custom implementation of the `ResourceLimitChecks`  are not restricted to use only data volume to limit the incoming messages. Custom implementations should be able to use any attributes (no. of messages, data volume, etc.) or combination of them. Giving a more generic method name for the message limit check in `ResourceLimitChecks` is sufficient for that purpose. The limit configurations are used only within the `ResourceLimitChecks` implementation and they need not be generic. _data-volume_ and _max-bytes_ sounds good.